physical/raft/raft.go

```diff
@@ -395,33 +400,71 @@ func NewRaftBackend(conf map[string]string, logger log.Logger) (physical.Backend
 		snap = raft.NewInmemSnapshotStore()
 	} else {
 		// Create the base raft path.
-		path := filepath.Join(path, raftState)
-		if err := EnsurePath(path, true); err != nil {
+		raftBasePath := filepath.Join(path, raftState)
+		if err := EnsurePath(raftBasePath, true); err != nil {
 			return nil, err
 		}
+		dbPath := filepath.Join(raftBasePath, "raft.db")
 
-		// Create the backend raft store for logs and stable storage.
-		dbPath := filepath.Join(path, "raft.db")
-		opts := boltOptions(dbPath)
-		raftOptions := raftboltdb.Options{
-			Path:        dbPath,
-			BoltOptions: opts,
+		// Check if they're opting in to raft-wal storage
+		useRaftWal := false
+		if walRaw, ok := conf[raftWalConfigKey]; ok {
+			useRaftWal, err = strconv.ParseBool(walRaw)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse %s config value %q as a boolean: %w", raftWalConfigKey, walRaw, err)
+			}
 		}
-		store, err := raftboltdb.New(raftOptions)
-		if err != nil {
-			return nil, err
+
+		// If the existing raft db exists from a previous use of BoltDB, warn about this and continue to use BoltDB
+		_, err := os.Stat(dbPath)
+		if useRaftWal && errors.Is(err, fs.ErrExist) {
```

Comment:
Does this work? Will `os.Stat` return a non-nil error in the case the file exists? my epectation is that it would just return a result and nil error in this case.

I think you need to check that err is nil here or if not nil is something _other_ than `fs.ErrNotExist`. It's kinda fiddly. This is how I ended up doing it in Consul: https://github.com/hashicorp/consul/blob/e235c8be3c67ed1389af017a76b29a8452b86453/agent/consul/server.go#L2141-L2153

Reply:
These things are a little hard to test too but since this is in a separate backend component I think it would be easier than in Consul to unit test this with real filesystem - configure to use WAL and then either intercept logs and assert on warning or poke inside later and assert on the raft config actually having boltDB still or something?

Reply:
Thanks! I stole your function and will use that.

Reply:
I'll make a note to write a test specifically to cover this case.